File: app/src/main/java/com/example/spelltracker/SpellDatabase.java
package com.example.spelltracker;

import android.content.Context;

import androidx.room.Database;
import androidx.room.Room;
import androidx.room.RoomDatabase;

/**
 * Singleton-обёртка над Room-базой {@code spells.db}.
 * Версия схемы: 1.
 */
@Database(entities = {Spell.class}, version = 1, exportSchema = false)
public abstract class SpellDatabase extends RoomDatabase {

    private static final String DB_NAME = "spells.db";
    private static volatile SpellDatabase INSTANCE;

    public abstract SpellDao spellDao();

    public static SpellDatabase get(Context context) {
        if (INSTANCE == null) {
            synchronized (SpellDatabase.class) {
                if (INSTANCE == null) {
                    INSTANCE = Room.databaseBuilder(
                                    context.getApplicationContext(),
                                    SpellDatabase.class,
                                    DB_NAME)
                            .build();
                }
            }
        }
        return INSTANCE;
    }
}

File: app/src/main/java/com/example/spelltracker/data/SpellDatabase.kt
package com.example.spelltracker.data

import android.content.Context
import androidx.room.Database
import androidx.room.Room
import androidx.room.RoomDatabase

/**
 * Room-база со справочником заклинаний.
 *
 * Версия 5 — добавили [Spell.subclassParents] (CSV English id родительских
 * классов для каждого подкласса). Миграция destructive: данные приходят из
 * assets/spells_normalized.json (см. generateSpellsDb в app/build.gradle.kts).
 *
 * `fallbackToDestructiveMigration` уничтожит старую БД при изменении
 * схемы; данные перечитаются из assets при следующем старте. Потеря
 * не страшна — это справочник.
 */
@Database(entities = [Spell::class], version = 5, exportSchema = false)
abstract class SpellDatabase : RoomDatabase() {

    abstract fun spellDao(): SpellDao

    companion object {
        @Volatile private var INSTANCE: SpellDatabase? = null

        fun get(context: Context): SpellDatabase = INSTANCE ?: synchronized(this) {
            INSTANCE ?: Room.databaseBuilder(
                context.applicationContext,
                SpellDatabase::class.java,
                "spells.db"
            )
                // Если на устройстве лежит spells.db от старой версии —
                // Room не сможет валидировать identity hash и упадёт.
                // Поскольку данные восстанавливаются из assets, безопаснее
                // дропнуть старую БД и пересоздать.
                .fallbackToDestructiveMigration()
                .build()
                .also { INSTANCE = it }
        }
    }
}
